### PR TITLE
Functional API: Segments.recursive_crawl

### PR DESCRIPTION
**Pull request by jpy-git**
### Brief summary of the change made

As discussed [here](https://github.com/sqlfluff/sqlfluff/pull/2364#discussion_r788731722) we should add recursive_crawl to the functional API.

### Are there any other side effects of this change that we should be aware of?
No

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: test/core/rules/functional/test_segments.py [hunk 1]
```diff
@@ -138,3 +139,25 @@ def test_segment_predicates_and():
...
+        tab_b.col_b
+    FROM cte
+    INNER JOIN tab_b;
+    """
+    linter = Linter()
+    parsed = linter.parse_string(sql)
+
+    functional_tree = segments.Segments(parsed.tree)
+
+    assert len(functional_tree.recursive_crawl("whitespace")) == 18
```
Comment (tunetheweb) [inline test/core/rules/functional/test_segments.py:161]: Seems a bit of a flaky test (and annoying to calc/update if test changes). Would the two test cases below suffice?

**Review by tunetheweb — APPROVED** [review]
LGTM. One minor comment.

File: test/core/rules/functional/test_segments.py (hunk 1 above)
Comment (jpy-git) [reply test/core/rules/functional/test_segments.py:161]: sounds good I just picked random segments

Comment (jpy-git) [reply test/core/rules/functional/test_segments.py]: updated now

**Pull request merged**